File: observer.py
class Subject:

    def __init__(self):
        self.observers = []
        
    def add_observer(self, observer):
        self.observers.append(observer)

    
    def remove_observer(self, observer):
        self.observers.remove(observer)

       
    def notify(self, data): 
        if(len(self.observers) > 0):
            for observer in self.observers:
                observer.update(data)
        else:
            print("Keine Observer registriert")

       
class Observer:

    def update():
        pass

        
class State(Subject):
    def __init__(self):
        super().__init__()
        self.state = []
    
    def add_person(self, person):
        self.state.append(person)
        self.notify(self.state)
    
    def get_state():
        return self.state


class PersonOutput(Observer):

    def update(self, state):
        print("Person ", state[-1]["name"], " Hinzugefügt")
        print(self)


class PersonCount(Observer):

    def update(self, state):
       return print("Anzahl der Personen: ", len(state))

class PersonList(Observer):
    def update(self, state):
        print("Liste der Personen: ")
        for person in state:
            print(person["name"])


state = State()
person_output = PersonOutput()
person_count  = PersonCount()
person_list = PersonList()

state.add_observer(person_output)
state.add_observer(person_count)
state.add_observer(person_list)
state.add_person({"name" : "Alvin"})
state.add_person({"name" : "Simon"})

File: observer.js
  class Subject{  //Subject
      constructor(){
        this.observers = [];
        }

      addObserver(observer){
      this.observers.push(observer); 
      }

      removeObserver(observer){
        this.observers = this.observers.filter(function(item){
          if(item !== observer)
          return item;
        });
      }
      
      notify(state){
        if (this.observers.length > 0){
          for(let observer of this.observers){
            observer.update(state);
          }
        }else{
          console.log("Keine Observer registriert")
        }
      }
    }


  class Observer{
    update(){}
  }


  class State extends Subject{
    constructor(){
      super();
      this.state = [];
    }

    addPerson(person){
      this.state.push(person);
      this.notify(this.state);
    }
    getState(){
      return this.state;
    }
  }


  class PersonOutput extends Observer{
    
    update(state){
      console.log("Person " + state[state.length -1].name +" Hinzugefügt");
    }
  }


  class PersonCount extends Observer{
    
    update(state){
      console.log("Anzahl der Personen: " + state.length)
      }
  }
  class PersonList extends Observer{

    update(state){
      console.log("Liste der Personen: ");
      for(let person of state)
        console.log(person.name)
    }
  }

const state = new State();
const personOutput = new PersonOutput();
const personCount = new PersonCount();
const personList = new PersonList();

state.addObserver(personOutput);
state.addObserver(personCount);
state.addObserver(personList);
state.addPerson({name: "Alvin"});
state.addPerson({name: "Simon"});
state.removeObserver(personCount);
state.addPerson({name: "Theodore"});



